test/test.js

```diff
@@ -474,6 +484,15 @@ describe('Page', function() {
       const result = await page.evaluate(() => window);
       expect(result).toBe(undefined);
     });
+    it('should fail for circular object', async({page, server}) => {
+      const result = await page.evaluate(() => {
+        const a = {};
+        const b = {a};
+        a.b = b;
+        return a;
+      });
+      expect(result).toBe(undefined);
```

Comment:
@aslushnikov Is there any reason on why this should return `undefined`. I think throwing an error would be more appropriate. Think of the consumer of this API. He will not realise his result is circular, or it might happen only in some occasions. Or maybe sometimes `undefined` is a legit output of his function. How will he be able to distinguish these cases from a bug? How will he be notified of bugs in his code?

Reply:
@Janpot the motivation behind this was to let people use `page.evaluate` safely without caring on the return value. Consider the following simple script:

```js
await page.evaluate(() => window.foo = window);
```

Technically, the `page.evaluate` should throw since the expression returns non-serializable value,
but most of the time its unintentional.

> How will they be able to distinguish these cases from a bug? How will they be notified of bugs in their code?

I filed https://github.com/GoogleChrome/puppeteer/issues/2079 with a solution that will hopefully address your concerns and satisfy our usecase.

Reply:
@aslushnikov How did you arrive at this conclusion: _but most of the time its unintentional._?
In my experience it's exactly the opposite, I've never had to evaluate and not care about the result.

Reply:
> How did you arrive at this conclusion: but most of the time its unintentional.?

@Janpot once we started throwing errors, we instantly had a bunch of feedback on both bugtracker and internally about the unwanted errors caused by one-liners in `page.evaluate`.

Reply:
Ok, I don't really agree with API design like this but I see you want to be sort of pragmatic, and I don't have arguments against that that would actually change your mind 🙂. I'll settle for
```js
async function evaluate (page, ...args) {
  const result = await page.evaluate(...args);
  if (result === puppeteer.UNSERIALIZABLE_VALUE) {
    throw new Error('Unserializable value');
  }
  retun result;
}
```
then.